crates/source_file/src/source_location.rs

```diff
@@ -0,0 +1,92 @@
+use std::fmt::Debug;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SourceLocation {
+    line_number: LineNumber,
+    line_offset: LineOffset,
+}
+
+impl SourceLocation {
+    pub fn new(line_number: LineNumber, line_offset: LineOffset) -> Self {
+        Self {
+            line_number,
+            line_offset,
+        }
+    }
+
+    pub fn line_number(&self) -> LineNumber {
+        self.line_number
+    }
+
+    pub fn line_offset(&self) -> LineOffset {
+        self.line_offset
+    }
+
+    pub fn into_fields(self) -> (LineNumber, LineOffset) {
+        (self.line_number, self.line_offset)
+    }
+}
+
+/// A 0-indexed line number
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct LineNumber(u32);
+
+impl From<u32> for LineNumber {
+    fn from(value: u32) -> Self {
+        LineNumber(value)
+    }
+}
+
+impl TryFrom<usize> for LineNumber {
+    type Error = std::num::TryFromIntError;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        Ok(LineNumber(u32::try_from(value)?))
+    }
+}
+
+impl From<LineNumber> for u32 {
+    fn from(value: LineNumber) -> Self {
+        value.0
+    }
+}
+
+impl From<LineNumber> for usize {
+    fn from(value: LineNumber) -> Self {
+        value.0 as usize
+    }
+}
+
+/// A 0-indexed offset into a line, represented as a number of code units under one of the
+/// three possible [LineOffsetEncoding]s
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct LineOffset {
+    raw: u32,
+    encoding: LineOffsetEncoding,
+}
```

Comment:
One other footgun / confusion point that I think I've nicely removed is that the _offset into the line_ is the _only_ thing that actually relies on the encoding, so the offset itself also owns the encoding.

I think this just really clarifies why the encoding is required, and how it gets used